awspropagator/src/main/java/software/amazon/opentelemetry/awspropagator/AwsCompositePropagator.java

```diff
@@ -0,0 +1,124 @@
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * A {@link TextMapPropagator} for use by AWS services or compatible ones such as localstack. As AWS
+ * is generic cloud infrastructure, rather than enforcing a specific injection format such as
+ * W3CTraceContext, it is intended to inject the same format as was extracted to allow compatibility
+ * with client applications that may not understand newer formats like W3C.
+ *
+ * <p>To allow rolling this propagator out to independent services, it also supports a mode that
+ * always injects in AWS format. This is intended for a transition period where most services will
+ * only understand that format.
+ */
+public final class AwsCompositePropagator implements TextMapPropagator {
+
+  // The currently supported propagators. This may grow or be made configurable in the future.
+  private static final List<TextMapPropagator> PROPAGATORS =
+      Arrays.asList(
+          W3CBaggagePropagator.getInstance(),
+          AwsXrayPropagator.getInstance(),
+          W3CTraceContextPropagator.getInstance(),
+          B3Propagator.injectingMultiHeaders());
+
+  /**
+   * Returns a {@link AwsCompositePropagator} which injects in the same format it extracted a trace
+   * context with. For example, if an incoming request contained a trace context in B3 format, an
+   * outgoing request will have B3 format injected.
+   */
+  public static AwsCompositePropagator injectingExtractedFormat() {
+    return new AwsCompositePropagator(true);
+  }
+
+  /**
+   * Returns a {@link AwsCompositePropagator} which injects in the AWS format (X-Amzn-Trace-Id). For
+   * example, if an incoming request contained a trace context in B3 format, an outgoing request
+   * will have AWS format injected.
+   */
+  public static AwsCompositePropagator injectingAwsFormat() {
+    return new AwsCompositePropagator(false);
+  }
+
+  private static final ContextKey<TextMapPropagator> EXTRACTED_PROPAGATOR =
+      ContextKey.named("extracted-propagator");
+
+  private final List<String> fields;
+  private final boolean injectExtractedFormat;
+
+  private AwsCompositePropagator(boolean injectExtractedFormat) {
+    this.injectExtractedFormat = injectExtractedFormat;
+
+    fields = PROPAGATORS.stream().flatMap(p -> p.fields().stream()).collect(Collectors.toList());
+  }
+
+  @Override
+  public Collection<String> fields() {
+    return fields;
+  }
+
+  @Override
+  public <C> void inject(Context context, C carrier, TextMapSetter<C> setter) {
+    if (injectExtractedFormat) {
+      TextMapPropagator extractedPropagator = context.get(EXTRACTED_PROPAGATOR);
+      if (extractedPropagator != null) {
+        extractedPropagator.inject(context, carrier, setter);
+        Baggage baggage = Baggage.fromContextOrNull(context);
+        if (baggage != null && extractedPropagator != AwsXrayPropagator.getInstance()) {
```

Comment:
So only the X-Ray propagator supports baggage as part of the trace header?

Reply:
The only other one I know of is jaeger propagator - if we add support for jaeger propagator than this could would need a tweak but for our current list it works well.